sdk/src/fee.rs

```diff
@@ -91,6 +103,25 @@ impl FeeStructure {
             1.0 // multiplier that has no effect
         };
 
+        (self
+            .calculate_fee_details(
+                message,
+                budget_limits,
+                include_loaded_account_data_size_in_fee,
+            )
+            .total_fee() as f64
+            * congestion_multiplier)
+            .round() as u64
+    }
+
+    /// Calculate fee details for `SanitizedMessage`
+    #[cfg(not(target_os = "solana"))]
+    pub fn calculate_fee_details(
+        &self,
+        message: &SanitizedMessage,
+        budget_limits: &FeeBudgetLimits,
+        include_loaded_account_data_size_in_fee: bool,
+    ) -> FeeDetails {
```

Comment:
I hate that such a small change is becoming so difficult, but it seems dangerous to introduce this while we still have the rounding, since `calculate_fee_details` and `calculate_fee` give different overall results. A call-site switching to `calculate_fee_details` could lead be a consensus-breaking change.

Seems like we almost need to get rid of congestion multiplier/round before we can safely introduce this.

Reply:
Thanks for being thorough on this. The multiplier and rounding is pia. Completely removing congestion multiplier breaks a lot of tests that require zero tx fee. It's probably not something can be done quickly. Removing rounding would need feature gate. 

Since adding FeeDetails is for Feature gate issue #34731, maybe can put this new pub function behind that gate, so call site can't switch to it without feature being activated first; tho don't want to make feature gate do more things than necessary. let me think about it.

Reply:
I think it's better to deal `congestion_multiplier` separately. The issue here that could break consensus is the rounding (#34982), we need to make sure `calculate_fee() -> u64` and `calculate_fee_details.total_fee() -> u64` always return same value.
  
Comment 637be2fdfe2422ea202db27da548b88b0ba8adb2 removes rounding by feature gate, then need to make sure #34982 is scheduled before #34731.